ENH: linalg: add batch support for functions that accept a single array

mdhaber opened the pull request:
#### Reference issue
Follow-up to gh-22127
Toward gh-21466

#### What does this implement/fix?
This adds support for n-dimensional batch input to `linalg` functions `diagsvd`, `inv`, `null_space`, `logm`, `sinm`, `cosm`, `tanm`, `sinhn`, `coshm`, `tanhm`, `sqrtm`, `funm`, `signm`, `fractional_matrix_power`, `pinv`, `pinvh`, `matrix_balance`, `bandwidth`, `svd`, `ldl`, `cholesky`, `polar`, `qr`, `rq`, `hessenberg`, `schur`, `orth`, `lu_factor`, `eig_banded`, and `eigvals_banded` because they all accept only one array input. (Please check me on this.)

#### Additional information
Even though functions like `cosm` are expressed in terms of `expm`, we might as well apply the decorator to them separately because `expm` just loops in Python, too. This is less work and we get the batch support documentation for free.

mdhaber committed:
ENH: linalg.diagsvd: add batch support

mdhaber committed:
ENH: linalg.inv: add batch support

mdhaber committed:
ENH: linalg.null_space: add batch support

mdhaber committed:
ENH: linalg.sqrtm: add batch support

mdhaber committed:
ENH: linalg.funm: add batch support

mdhaber committed:
ENH: linalg.signm: add batch support

mdhaber committed:
ENH: linalg.fractional_matrix_power: add batch support

mdhaber committed:
ENH: linalg: add batch support to several functions

j-bowhay commented:
Xref https://github.com/scipy/scipy/pull/21446#issuecomment-2311841148 it might be worth checking if the decorator is needed for some of these functions

j-bowhay commented on scipy/linalg/tests/test_batch.py:
```diff
@@ -21,36 +29,75 @@ def get_nearly_hermitian(shape, dtype, atol, rng):
+
+    @pytest.mark.parametrize('dtype', floating)
+    def test_logm(self, dtype, rng):
+        with warnings.catch_warnings():  # "result not accurate"
```
Looks like the ci failures are related to not using `numpy.testing.suppress_warnings`

mdhaber replied:
<sigh>

mdhaber replied:
Actually it turns out that increasing the diagonal elements was enough to avoid these warnings, too, so I was able to just remove the filter.

mdhaber committed:
MAINT: linalg.logm: remove warnings filter from test

mdhaber commented on scipy/linalg/tests/test_batch.py:
```diff
@@ -1,3 +1,5 @@
+import warnings
+
```
```suggestion
```

mdhaber replied:
Will commit with `[lint only]` if the rest passes.

mdhaber commented on scipy/linalg/tests/test_batch.py:
```diff
@@ -21,36 +29,73 @@ def get_nearly_hermitian(shape, dtype, atol, rng):
 
 class TestMatrixInScalarOut:
```
```suggestion
class TestOneInOneOut:
```

mdhaber committed:
ENH: linalg.pinv/pinvh: add batch support

mdhaber committed:
ENH: linalg.matrix_balance: add batch support

j-bowhay reviewed (changes requested):
Maybe also worth passing the batched input a keyword argument in the tests?

mdhaber commented:
Good idea, or if it's ok with you, I'll just inspect for the name of the argument? I didn't in the initial PR to keep things "simple", but I'll show that the overhead is negligible, and clearly it's less prone to error.

mdhaber committed:
TST: linalg: test passing arrays by keyword

mdhaber committed:
ENH: linalg.bandwidth: add batch support

mdhaber committed:
ENH: linalg.ldl: add batch support

mdhaber committed:
ENH: linalg.svd: add batch support

mdhaber committed:
ENH: linalg.cholesky: add batch support

mdhaber committed:
ENH: linalg.polar: add batch support

mdhaber committed:
ENH: linalg: add batch support to qr, rq, schur, hessenberg

mdhaber committed:
MAINT: linalg.clarkson_woodruff_transform: add batch support

mdhaber committed:
ENH: linalg.orth/lu_factor: add batch support

mdhaber committed:
ENH: linalg.eig_banded/eigvals_banded: add batch support

ilayn commented:
Not to block any of these nice additions;

For the matrix functions, I have some time to finish up `sqrtm` which probably makes `logm` also reachable for batch handling. I can probably make them behave like `expm`. Would that be an anti-pattern for the decorator usage you are adding here?

j-bowhay commented:
> Not to block any of these nice additions;
> 
> For the matrix functions, I have some time to finish up `sqrtm` which probably makes `logm` also reachable for batch handling. I can probably make them behave like `expm`. Would that be an anti-pattern for the decorator usage you are adding here?

I think the idea is for any of these is, if batching in the compiled code materialises then the decorator can be removed. Plus this means there's already a test for batching there ready and waiting:)

mdhaber committed:
TST: linalg.bandwidth: no more error with 3D input

mdhaber committed:
MAINT: linalg.cwt: remove batch support to avoid issues with sparse input

mdhaber commented:
Agreed - just adding that something would need to be done for the documentation, since the decorator adds the standard note right now. 
Also @ilayn it looks like `expm` is just looping in Python.
https://github.com/scipy/scipy/blob/acf6736a2325b71a69a2189125a5c8ceaea312ee/scipy/linalg/_matfuncs.py#L316-L318
Is that what you had in mind? I would not add more of that when the decorator does the same thing.

mdhaber commented:
One of the CI failures was a test that wanted to see an error with 3D input, so I removed that. The other involved CWT with sparse input. The decorator doesn't consider that right now, so I just removed it from CWT.

ilayn commented:
That python loop is not affecting since the `expm` computation typically dominating the runtime cost. So I decided to leave it on the Python side instead of dealing with memory addresses on the C side for slicing. But for very light wrappers around LAPACK calls, it can be slow to loop in Python. 

I just wanted to know if we can remove the decorator and still be consistent in case we do the batching internally. I guess we can with a bit of extra work.

j-bowhay reviewed (approved):
Thanks!
